Convert API to use std::shared_ptr

de-vri-es opened the pull request:
This PR updates the API to use `std::shared_ptr` for moveit types instead of `boost::shared_ptr`.

I'm not sure what the opinions are on this topic, but this PR at the least provides a place to discuss this specific API change. But do note that other libraries are doing the same thing right about now (like `fcl` and `urdfdom`).

Technically, switching from `boost::enable_shared_from_this` to `std::enable_shared_from_this` is not 100% API compatible, since `weak_from_this` is only added in C++17. This is very unlikely to cause problems though.

One more point: `pluginlib` still returns `boost::shared_ptr` and using `createUnmanagedInstance` is discouraged. Instead I opted to convert the pointers to `std::` instead. That method is not 100% safe when `weak_ptr` are involved in the general case, but it is in this case because nobody else has a copy to the original `boost::shared_ptr`.

Because the conversion isn't safe in general I didn't want to put it in a header (the header would have to be exported or duplicated across 5 packages). Instead I opted now to duplicate the function in an anonymous namespace in 5 source files. I've also got a PR open with `class_loader` to add support for `std::` smart pointers (unique and shared), so when that lands it makes sense to adapt to that interface. In the mean time I didn't know a nicer solution than the current in this PR.

davetcoleman commented on moveit_core/collision_detection/include/moveit/collision_detection/world.h:
```diff
@@ -44,12 +44,15 @@
-MOVEIT_CLASS_FORWARD(Shape);
+class Shape;
+typedef boost::shared_ptr<Shape> ShapePtr;
+typedef boost::shared_ptr<const Shape> ShapeConstPtr;
```
Just to make sure I understand the logic here - we don't want to use std::shared_ptr for ``Shape`` because its part of geometric_shapes, which is part of ``desktop-full``, which is not suppose to have std in its public headers yet, correct?

https://github.com/ros-planning/geometric_shapes/blob/kinetic-devel/include/geometric_shapes/shapes.h#L273

de-vri-es replied:
I agree, it would be nice to have `std::shared_ptr` there. The previous deviation from REP-003 was out of necessity though. On the other hand, `urdfdom` also switched. And while it may not be real  ROS package that is more or less a technicality.

But yeah, as long as the `geometric_shape` API uses `boost::shared_ptr` we can't switch them here.

davetcoleman replied:
How about a comment for ``ROS-L`` saying we'll switch to C++11 then?

davetcoleman commented on moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:
```diff
@@ -72,6 +72,19 @@ bool checkTimeout(ros::Time &t, double timeout, bool force = false)
+{
+  /**
+   * @brief Convert a boost::shared_ptr to an std::shared_ptr as good as possible.
+   * This is a dangerous conversion that only works in specific cases. Do not use blindly!
```
great comments

davetcoleman commented on moveit_plugins/moveit_ros_control_interface/src/controller_manager_plugin.cpp:
```diff
@@ -72,6 +72,19 @@ bool checkTimeout(ros::Time &t, double timeout, bool force = false)
+  {
+    return std::shared_ptr<T>(ptr.get(), [ptr](T*) mutable {ptr.reset();});
+  }
+}
```
will all of these ``convert_shared_ptr`` go away when https://github.com/ros/class_loader/pull/38 is released?

de-vri-es replied:
Yes, definitely. The sooner they dissapear the better.

davetcoleman replied:
lets open an issue as a reminder to remove them

de-vri-es replied:
Will do if (when?) we agree on an intent to merge.

de-vri-es replied:
ros/class_loader#38 is merged, ros/pluginlib#42 is the last missing piece. Then the only thing remaining for deleting these functions is waiting for their releases.

de-vri-es replied:
The pluginlib PR is also merged, so just waiting for the release now.

davetcoleman replied:
Look like https://github.com/ros/pluginlib/pull/42 is released now

de-vri-es replied:
Indeed, I'll remove these functions soon (tm). Possibly this evening, but it's a bit busy at work right now.

v4hn replied:
yay! so this should be removed again before we merge the request.
Awesome! Thanks for pushing this through the different layers @de-vri-es

davetcoleman commented:
+1 but other maintainers should get a chance to review and discuss this big change
@v4hn @rhaschke @mikeferguson

davetcoleman commented:
I would like to merge this asap so we have momentum for the kinetic release

v4hn commented on moveit_core/robot_model/include/moveit/robot_model/link_model.h:
```diff
@@ -41,13 +41,16 @@
-MOVEIT_CLASS_FORWARD(Shape);
+class Shape;
+typedef boost::shared_ptr<Shape> ShapePtr;
+typedef boost::shared_ptr<const Shape> ShapeConstPtr;
```
gosh, how many of these cases are there? Why is this not centralized?

de-vri-es replied:
I was a bit hesitant to add a macro for declaring `boost::shared_ptr` if they're going to turn in to `std::shared_ptrs` in the not-too-far future. Having macros like that lying around could be seen as an invitation to use them.

How about adding these typedefs in `geometric_shapes` (and possibly a new header `geometric_shapes/forward.h` to avoid including complex header), and until those are available keeping them explicit like this?

davetcoleman replied:
+1, though I think you could make the switch now since kinetic is not released yet and Travis builds ``geometric_shapes`` from source for this PR

v4hn replied:
Ah, these forward delarations where added by @rhaschke in 13d2e622658c995e140f85a08891ea65ba8d7dd5 to reduce includes of geometric_shapes.
I don't like the idea to introduce a header there that only forward declares all the types.
@rhaschke is this change really worth the overhead to maintain the boost-shared_ptrs here?

On the other hand I agree with @davetcoleman. I don't think it's any good to have boost- *and* std- shared_ptr around in `geometric_shapes` in kinetic. So I would actually propose to fully move to `std::shared_ptr` there. This would also simplify the forward declarations here.

de-vri-es replied:
I like the idea of that. I think I'll have some questions but I'll open a PR there for those (or an issue if the questions are blocking).

de-vri-es replied:
Hmm, changing the type of `geometric_shapes::ShapePtr` is not really a nice solution. Best is if `geometric_shapes` dictates the type of `shared_ptr`, but that is not going to happen without including a `geometric_shapes` header.

So that leaves pretty much just two options:
* Add a forward declarations header to `geometric_shapes`
* Revert the largest part of 13d2e622658c995e140f85a08891ea65ba8d7dd5

@v4hn, @rhaschke: Thoughts?

v4hn replied:
I didn't think of it as a "solution to this problem". More as "this problem arises because the previous change that made `geometric_shapes` use c++11 shared_ptr was inconsistent.
And if we want to switch MoveIt! to std::shared_ptr (which is what this request is about), this should imho include geometric_shapes. Incidentally, I assumed *all* shared_ptr references where switched to c++11 when we discussed the issue with FCL and octomap.

de-vri-es replied:
Nevertheless, ros-planning/geometric_shapes#57 and using `MOVEIT_CLASS_FORWARD` here again will also solve the practical issue.

That would basically mean dropping one commit from this PR:
* Don't use MOVEIT_CLASS_FORWARD for non-moveit classes.

de-vri-es replied:
I didn't seem your last comment when replying. But fair enough, it does make sense to switch `geometric_shapes` and `moveit` at the same time. It could get very confusing otherwise.

And in that light it may be acceptable to use `MOVEIT_CLASS_FORWARD` for `geometric_shapes` types. It feels a bit weird to define typedefs using a moveit convention in the namespace of another library, but I suppose `geometric_shapes` and moveit are related enough.

Shall I just drop the relevant commit then, and revert to using `MOVEIT_CLASS_FORWARD`?

v4hn replied:
let's see what comes out of your new request for geometric_shapes.

davetcoleman replied:
I was assuming we were not switching these because it would violate the ROS policy of no std::shared_ptrs in the public headers of packages in ros-desktop-full. But then again, we already had to break that rule for FCL and Octomap, so it doesn't make a difference going all the way now.

v4hn commented on moveit_experimental/collision_distance_field/src/collision_world_distance_field.cpp:
```diff
@@ -37,7 +37,6 @@
 #include <moveit/collision_distance_field/collision_world_distance_field.h>
 #include <moveit/collision_distance_field/collision_common_distance_field.h>
 #include <moveit/distance_field/propagation_distance_field.h>
-#include <boost/make_shared.hpp>
```
this is missing `<memory>`

de-vri-es replied:
Same as previous, so will also add.

de-vri-es replied:
`#include <memory>` added

v4hn commented on moveit_plugins/moveit_ros_control_interface/src/joint_trajectory_controller_plugin.cpp:
```diff
@@ -51,7 +51,7 @@ class JointTrajectoryControllerAllocator : public ControllerHandleAllocator
                                                                      const std::vector<std::string> &resources)
   {
-    return boost::make_shared<moveit_simple_controller_manager::FollowJointTrajectoryControllerHandle>(
+    return std::make_shared<moveit_simple_controller_manager::FollowJointTrajectoryControllerHandle>(
```
`<memory>` is not included here, but `<boost/shared_ptr.hpp>` still is. Is the latter still required here?

de-vri-es replied:
Good catch. Indeed, boost header removed, `#include <memory>` added

v4hn commented on moveit_ros/manipulation/pick_place/include/moveit/pick_place/pick_place.h:
```diff
@@ -45,7 +45,6 @@
 #include <moveit_msgs/PickupAction.h>
 #include <moveit_msgs/PlaceAction.h>
 #include <boost/noncopyable.hpp>
-#include <boost/enable_shared_from_this.hpp>
```
`<memory>` should be included for `std::enable_shared_from_this`

de-vri-es replied:
`#include <memory>` added

v4hn commented on moveit_ros/perception/semantic_world/include/moveit/semantic_world/semantic_world.h:
```diff
@@ -42,12 +42,15 @@
-MOVEIT_CLASS_FORWARD(Shape);
+class Shape;
+typedef boost::shared_ptr<Shape> ShapePtr;
+typedef boost::shared_ptr<const Shape> ShapeConstPtr;
```
and yet another case that should have a comment saying this should be changed in lunar.

de-vri-es commented on moveit_core/collision_detection/include/moveit/collision_detection/world_diff.h:
```diff
@@ -124,7 +123,7 @@ namespace collision_detection
 
     /* used to unregister the notifier */
-    boost::weak_ptr<World> world_;
+    std::weak_ptr<World> world_;
```
We had a short discussion at #206. Nobody seemed very interested (including me), but if you think that is the right approach you can reopen the PR.

davetcoleman replied:
-1 to that macro

de-vri-es commented on moveit_core/collision_detection/include/moveit/collision_detection/world_diff.h:
```diff
@@ -39,7 +39,6 @@
 
 #include <moveit/collision_detection/world.h>
 #include <moveit/macros/class_forward.h>
-#include <boost/weak_ptr.hpp>
```
Well, you can't make `weak_ptrs` without `shared_ptrs`. Since we have `std::shared_ptr` typedefs available from a macro in `#include <moveit/macros/declare_ptr.h>`, that will have pulled in `#include <memory>`.

But I am a fan of the rule "if you use a symbol from a header, include it", so I'll add it.

de-vri-es commented on moveit_core/planning_scene/include/moveit/planning_scene/planning_scene.h:
```diff
@@ -51,7 +51,6 @@
 #include <moveit_msgs/PlanningSceneComponents.h>
-#include <boost/enable_shared_from_this.hpp>
 #include <boost/noncopyable.hpp>
 #include <boost/shared_ptr.hpp>
```
Yeah, line 96 and 887 still use it for non-moveit types.

v4hn replied:
yes, but now thant you use the Ptr typedef they don't mention the symbol anymore. So this can be removed imo

de-vri-es replied:
Good point, will check.

de-vri-es commented:
hmm, seems something went wrong during a rebase

de-vri-es commented:
Ah, the new `BenchmarkExecutor` wasn't using the typedefs. Added a commit for that. Looks like that commit doesn't need to be cherry picked to Indigo and Jade either.

davetcoleman commented:
correct, ``BenchmarkExecutor`` only exists in Kinetic and will not be backported

de-vri-es commented:
Looks like the tests are failing because the version of pluginlib is too old. Probably needs to wait for a docker image rebuild?

Bit off-topic: apparently the `version_gte` tags in `package.xml` are just informative? Couldn't find any documentation on what they actually do, but it seems catkin and rosdep are ignoring them.

gavanderhoorn commented:
@de-vri-es wrote:
> Bit off-topic: apparently the `version_gte` tags in `package.xml` are just informative? Couldn't find any documentation on what they actually do, but it seems catkin and rosdep are ignoring them.

This is documented in REP-140, but you're right that `rosdep` doesn't use that information. See also ros-infrastructure/rosdep#325.

Bloom translates those into the appropriate key:value pairs in Debian control and RPM spec files.

davetcoleman commented:
I'm looking into why the [shadow-fixed](https://hub.docker.com/r/moveit/moveit/builds/bgymaxwavsuztewbmrchj2h/) CI is not building on Dockerhub

davetcoleman commented:
Docker error is described here: https://github.com/ros/ros_comm/issues/904

de-vri-es commented:
@gavanderhoorn: Thanks for the information. Not sure how I missed that.

Update regarding this PR:
* Dropped the commit that undid `MOVEIT_CLASS_FORWARD(Shape)`.
* Switched out a `boost::dynamic_ptr_cast` for the `std` equivalent.
* Replaced `boost::scoped_ptr` with `std::unique_ptr`.

I \*think\* this addresses all the remaining review points, but correct me if I'm wrong.

Only the first change affects the MoveIt API. The other two are implementation changes. The last point wasn't technically needed, but I think it makes sense to remove all boost smart pointers in one go.

Locally this builds (with ros-planning/geometric_shapes#57), including the tests. I've never been able to run all tests successfully on my system though (probably due to python3 issues which I intend to investigate later). I don't expect any problems though, considering the interface and semantics of `boost::` and `std::shared_ptr` are virtually identical (so the compiler should have caught any problems).

de-vri-es commented:
Actually, there are still `boost::shared_ptrs` left. It is possible that many of these can also be changed, but it needs to be verified that the library owning the type doesn't use `boost::shared_ptr` in their own API.

There are a number of cases in `moveit_core` and `moveit_ros`. Mostly these are `shared_ptrs` to tf, srdf, and urdf types. The real tricky bit here is knowing if it conflicts with other libraries besides the one owning the type. Many of these `shared_ptrs` might not even be part of the MoveIt API though, in which case it is safe. Should I look into these cases in more detail?

v4hn commented:
For tf I would stick with `boost::shared_ptr` for kinetic as there are interfaces in other packages that want to take ownership of a tf-listener and expect the boost ptr.

I'm unsure about the others. `grep -r boost::shared_ptr --include *.h | grep -v tf::Transform` basically lists `moveit_experimental/collision_distance_field` that still uses boost ptrs internally and a small number of diverse cases that need individual review.

But yes, I would welcome changing these cases (and also the non-API cases) too.

de-vri-es commented:
The last commit converts a number of internal `shared_ptr` to `std::shared_ptr`. I believe they were all safe to be converted.

Ignoring `boost::shared_ptr<tf::...>` and everything in moveit_experimental, chomp and sbpl, there is only one explicit `boost::shared_ptr<urdf::Model>` left in the setup assistant. But I think that is best left until ros/robot_model#153 is available.

v4hn commented:
On Sat, Sep 24, 2016 at 02:09:56PM -0700, Maarten de Vries wrote:
> The last commit converts a number of internal `shared_ptr` to `std::shared_ptr`. I believe they were all safe to be converted.

LGTM

> Ignoring `boost::shared_ptr<tf::...>` and everything in moveit_experimental, chomp and sbpl

Recently #155 has been merged. I didn't look into it, but if it does "resurrect chomp",
the code there should be switched too.

de-vri-es commented:
> Recently #155 has been merged. I didn't look into it, but if it does "resurrect chomp",
> the code there should be switched too.

There are no `package.xml` files, so I assumed work isn't finished yet.

Rebased on latest kinetic-devel.

davetcoleman commented:
I re-setup a [temporary Travis](https://travis-ci.org/davetcoleman/moveit_kinetic_cpp11/builds/162628953) build for the two C++11 pull requests in [this repo](https://github.com/davetcoleman/moveit_kinetic_cpp11). Once it passes I'm ready to merge this.

davetcoleman commented:
pluginlib didn't seem to release properly: https://github.com/ros/pluginlib/issues/43

de-vri-es commented:
I've got a few commits ready for #249 when it gets merged.

davetcoleman commented:
Could you review and +1 #249? I shouldn't merge it since I added some small fixes at the end for the kinetic branch.

de-vri-es commented:
Rebased on latest kinetic-devel again, now with collision_distance_field and chomp.

davetcoleman commented:
Need https://github.com/ros-planning/moveit/pull/250 to be merged before test build https://github.com/davetcoleman/moveit_kinetic_cpp11 passes, indicating this PR is good

davetcoleman commented:
@de-vri-es can you rebase this PR with the head of kinetic-devel? The build is failing because of the now fixed fcl dependency

de-vri-es commented:
Rebased. Note that there are two commits that we probably want to cherry-pick to I/J (using commit messages in case of future rebases):
* Use srdf::ModelPtr typedefs.
* Use shared_ptr typedefs in collision_distance_field and chomp.

de-vri-es commented:
The travis test passes \o/
https://travis-ci.org/davetcoleman/moveit_kinetic_cpp11

davetcoleman commented:
darn there is a merge conflict now, and i just merged https://github.com/ros-planning/geometric_shapes/pull/57

can you fix asap?

de-vri-es commented:
On it, rebase worked automatically, but I'd like to compile before pushing to be sure (I find automatic merges scary). Should be about 5 to 10 minutes.

de-vri-es commented:
Compiled fine locally, pushed.

davetcoleman commented:
:clap: C++11!

de-vri-es commented:
\o/

v4hn commented:
I fully agree with this change, but to point this out:

This change entails that *all* packages that compile with a `moveit_core` dependency in Xenial (the major ubuntu distribution supported by ROS kinetic) *have to explicitly specify* a standard >= c++11 in the build scripts to compile their packages. This is because Xenial ships gcc 5.3.1 which does not yet default to c++11.
At the moment this breaks our tutorials and packages in moveit_pr2 for kinetic (requests in progress).
